Analytical_Engine/Create/Graph/Graph.cs

```diff
@@ -129,6 +128,12 @@ public static Graph Graph<T>(List<ICurve> connectingCurves, T prototypeEntity, L
             if (entities == null)
                 entities = new List<IElement0D>();
 
+            if (!entities.All(e => e is IBHoMObject))
```

Comment:
Given the incoming `entities` is already a `List<IBHoMObject>` I'm not sure what benefit this if statement will have.

This feels a little bit like trying to put a plaster on a flesh wound - from reading through the issue there is an issue with certain object types and invalid casts, but it feels like this PR is solving a specific test case for the `Fitting` objects, but without considering the remainder of the algorithm(s) or uses.

Might be best to have a call on this and look to leverage some resourcing mechanisms cause I think this requires a more detailed review/overhaul than one PR can sufficiently patch?

Reply:
IElement0D never interfaces with nor inherits from BHoMObject or IBHoMObject....

Reply:
Which would suggest no `IElement0D` could ever end up in this list unless the object itself also implements `IBHoMObject` - but this check is still fairly redundant given the input object is of type `List<IBHoMObject>` already?

Reply:
https://github.com/BHoM/BHoM/blob/3264411efd671aeb808f79ab6d29d7cc4f7ba3cb/Geometry_oM/Vector/Point.cs#L30

Not a BHoMObject. 

Is IElement0D

Reply:
Our Anemic Domain Model, with use of extension methods only, allows for the problem where since interfaces cannot enforce properties or methods as a contract, that it is entirely possible that an object may satisfies the polymorphic relationship in terms of "is a" relationship, but does make it entirely possible that the desired contract of props and methods is not satisfied.

Reply:
Aye, but [this line](https://github.com/BHoM/BHoM_Engine/blob/develop/Analytical_Engine/Create/Graph/Graph.cs#L75) shows the input `entities` is of type `List<IBHoMObject>`, which means you wouldn't be able to have a `Point` in the list in the first place. Therefore, this `if` statement checking whether all the objects in list `entities` are `IBHoMObject`s is redundant because you couldn't get a Point object in the list to start with? E.G:

![image](https://github.com/BHoM/BHoM_Engine/assets/18049174/cf5f50b5-9a30-40ef-9075-695059964ece)

Bit small screenshot, apologies, but because `Point` is not an `IBHoMObject`, we can't add it to collections of `IBHoMObject` which the input of this method is. If the input was `object` or `IObject` or `IElement0D` then I think the if statement would be needed 😄

Reply:
But taking a step back, it would be good to know why the Guid is so important, which is why the object needs to be a BHoMObject as in lines here: https://github.com/BHoM/BHoM_Engine/blob/eb756492dbf5274f069bdcbbcac9641c6e19107b/Analytical_Engine/Create/Graph/Graph.cs#L140-L145

This is the root issue maybe that I have given too much weight to

Reply:
Agreed - it looks like the `Guid` was being used to uniquely identify objects when they might be more complex than just 3 vertices (as a point is), but given `BHoM_Guid` cannot be guaranteed to be the same every time, matching based on that is usually a recipe for disaster.

I would encourage us to consider the `Relation` object and maybe refactor it if needed. I think we could refactor it do be `IElement0D` for the Source and Target and use the whole object rather than just the `Guid` - though maybe the `Guid` was used for speed because some `IElement0D` objects can be data heavy when passing around/using within the system.